src/lib/mdns/minimal/ResponseSender.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ResponseSender.h"
+
+#include "QueryReplyFilter.h"
+
+#include <support/ReturnMacros.h>
+
+#define RETURN_IF_ERROR(err)                                                                                                       \
+    do                                                                                                                             \
+    {                                                                                                                              \
+        if (err)                                                                                                                   \
+        {                                                                                                                          \
+            return;                                                                                                                \
+        }                                                                                                                          \
+    } while (false)
+
+namespace mdns {
+namespace Minimal {
+
+namespace {
+
+constexpr uint16_t kMdnsStandardPort = 5353;
+constexpr uint16_t kPacketSizeBytes  = 512;
+
+} // namespace
+
+CHIP_ERROR ResponseSender::Respond(const QueryData & query, const chip::Inet::IPPacketInfo * querySource)
+{
+    mSendError = CHIP_NO_ERROR;
+
+    mCurrentSource = querySource;
+    mSendUnicast   = query.GetUnicastAnswer();
+
+    mResponder->ResetAdditionals();
+
+    mCurrentResourceType = ResourceType::kAnswer; // direct answer
+    QueryReplyFilter filter(query);
+    for (auto it = mResponder->begin(); it != mResponder->end(); it++)
+    {
+        Responder * responder = it->responder;
+
+        if (!filter.SendAnswer(responder->GetQType(), responder->GetQClass(), responder->GetQName()))
+        {
+            continue;
+        }
+
+        responder->AddAllResponses(querySource, this);
+        ReturnErrorOnFailure(mSendError);
+
+        mResponder->MarkAdditionalReplyesFor(it);
+    }
+
+    mCurrentResourceType = ResourceType::kAdditional; // Additional parts
+    filter.SetIgnoreNameMatch(true);
```

Comment:
This shows rough edges when I was refactoring and unsure if to create a new filter or not.

The intent is to 'send additionals' even if name differs (it always does, like PTR qname is not the same as the one that we are pointing to). I still want to respect the QClass/QType filter if any exists though, hence the check.

I think it could be refactored, however I would prefer to do it later. The whole code could use a few passes, however I would like to defer those for later as it is easier to review a stand-alone movement/renaming of things than one large PR that keeps changing.